File: frontend/src/api/datasets.ts
import { del, get, post } from "./client";

export interface FieldMapping {
  messages_key?: string;
  chosen_key?: string;
  rejected_key?: string;
  content_key?: string;
  role_key?: string;
}

const fieldMappingPayload = (m?: FieldMapping | null) => (m ? { field_mapping: m } : {});

export function listJsonl(path: string = "") {
  return get<{ files: { name: string; path: string }[] }>("/api/datasets/list", { path });
}

export function getRecords(path: string, file: string, offset = 0, limit = 50) {
  return get<{ records: Record<string, unknown>[]; total: number; offset: number; limit: number }>(
    "/api/datasets/records",
    { path, file, offset, limit }
  );
}

export function getRecord(path: string, file: string, index: number) {
  return get<{ index: number; data: Record<string, unknown> }>("/api/datasets/record", { path, file, index });
}

export interface TokenCountResult {
  messages_count: number;
  chosen_count: number;
  rejected_count: number;
  messages_plus_chosen: number;
  messages_plus_rejected: number;
  per_message_tokens: number[][];
}

export function tokenCount(
  path: string,
  file: string,
  index: number,
  model: string,
  fieldMapping?: FieldMapping | null
) {
  return post<TokenCountResult>("/api/datasets/token-count", {
    path,
    file,
    index,
    model,
    ...fieldMappingPayload(fieldMapping),
  });
}

export interface TokenStatsResult {
  mean?: number;
  min?: number;
  max?: number;
  n?: number;
  std?: number;
  median?: number;
  p25?: number;
  p75?: number;
  p90?: number;
  histogram?: { bucket_edges: number[]; counts: number[] };
  chosen_wise?: TokenStatsResult;
  rejected_wise?: TokenStatsResult;
}

export function tokenStats(
  path: string,
  file: string,
  model: string,
  scope: string,
  fieldMapping?: FieldMapping | null
) {
  return post<TokenStatsResult>("/api/datasets/token-stats", {
    path,
    file,
    model,
    scope,
    ...fieldMappingPayload(fieldMapping),
  });
}

export function ngram(
  path: string,
  file: string,
  n: number,
  min_count: number,
  min_length: number,
  scope: string,
  unit: "char" | "word" = "char",
  fieldMapping?: FieldMapping | null
) {
  return post<{ items: { gram: string; count: number }[] }>("/api/datasets/ngram", {
    path,
    file,
    n,
    min_count,
    min_length,
    scope,
    unit,
    ...fieldMappingPayload(fieldMapping),
  });
}

export interface StringSearchResult {
  total_occurrences: number;
  records_with_match: number;
  per_record: { index: number; count: number }[];
  mean_per_record?: number;
  min_per_record?: number;
  max_per_record?: number;
  std_per_record?: number;
}

export function stringSearch(
  path: string,
  file: string,
  query: string,
  scope: string,
  fieldMapping?: FieldMapping | null
) {
  return post<StringSearchResult>("/api/datasets/string-search", {
    path,
    file,
    query,
    scope,
    ...fieldMappingPayload(fieldMapping),
  });
}

export interface RunScriptResult {
  stdout: string;
  stderr: string;
  error?: string;
}

export function runScript(
  path: string,
  file: string,
  script: string,
  fieldMapping?: FieldMapping | null
) {
  return post<RunScriptResult>("/api/datasets/run-script", {
    path,
    file,
    script,
    ...fieldMappingPayload(fieldMapping),
  });
}

export interface SavedScriptInfo {
  id: string;
  name: string;
  description: string;
}

export function listScripts() {
  return get<{ scripts: SavedScriptInfo[] }>("/api/datasets/scripts");
}

export function getScript(scriptId: string) {
  return get<{ id: string; body: string }>(`/api/datasets/script/${encodeURIComponent(scriptId)}`);
}

export function saveScript(scriptId: string, body: string) {
  return post<{ ok: boolean }>("/api/datasets/script/save", { id: scriptId, body });
}

export function deleteScript(scriptId: string) {
  return del<{ deleted: boolean }>(`/api/datasets/script/${encodeURIComponent(scriptId)}`);
}

File: backend/api/datasets.py
"""Dataset API."""
from fastapi import APIRouter, HTTPException

from core.exceptions import PathOutsideWorkspaceError
from models.dataset import NgramRequest, RunScriptRequest, SaveScriptBody, StringSearchRequest, TokenCountRequest, TokenStatsRequest
from services import chat_stats_service, dataset_service, script_service, token_service

router = APIRouter(prefix="/api/datasets", tags=["datasets"])


@router.get("/list")
def list_jsonl(path: str = ""):
    try:
        items = dataset_service.list_jsonl(path)
        return {"files": items}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.get("/records")
def get_records(path: str, file: str, offset: int = 0, limit: int = 50):
    try:
        records, total = dataset_service.get_records(path, file, offset, limit)
        return {"records": records, "total": total, "offset": offset, "limit": limit}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.get("/record")
def get_record(path: str, file: str, index: int):
    try:
        record = dataset_service.get_record(path, file, index)
        if record is None:
            raise HTTPException(404, "Record not found")
        return {"index": index, "data": record}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


def _mapping_dict(m):
    if m is None:
        return None
    return m.model_dump() if hasattr(m, "model_dump") else m.dict()


@router.post("/token-count", summary="Token count for one record")
def post_token_count(body: TokenCountRequest):
    """Return token counts for a single record: messages, chosen, rejected, sums, and per-message token list."""
    try:
        result = token_service.token_count_record(
            body.path, body.file, body.index, body.model, field_mapping=_mapping_dict(body.field_mapping)
        )
        if result is None:
            raise HTTPException(404, "Record not found")
        return result
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except RuntimeError as e:
        raise HTTPException(503, str(e))


@router.post("/token-stats", summary="Token stats for full file")
def post_token_stats(body: TokenStatsRequest):
    """Return mean/min/max and histogram for token counts over the file. Scope: chosen_wise, rejected_wise, or both."""
    try:
        return token_service.token_stats_file(
            body.path, body.file, body.model, body.scope, field_mapping=_mapping_dict(body.field_mapping)
        )
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
    except RuntimeError as e:
        raise HTTPException(503, str(e))


@router.post("/ngram", summary="N-gram screening")
def post_ngram(body: NgramRequest):
    """Return n-gram frequency list. Scope: messages, chosen, rejected, all. Unit: char or word."""
    try:
        items = chat_stats_service.ngram(
            body.path,
            body.file,
            body.n,
            body.min_count,
            body.min_length,
            body.scope,
            body.unit,
            field_mapping=_mapping_dict(body.field_mapping),
        )
        return {"items": items}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.post("/string-search", summary="String search and count")
def post_string_search(body: StringSearchRequest):
    """Return total occurrences, records with match, and per-record counts. Scope: chosen_wise, rejected_wise, whole."""
    try:
        return chat_stats_service.string_search(
            body.path, body.file, body.query, body.scope, field_mapping=_mapping_dict(body.field_mapping)
        )
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.post("/run-script", summary="Run user script with workspace API")
def post_run_script(body: RunScriptRequest):
    """Execute script with workspace (get_record, get_all_records, get_field_as_list, get_text_list, run_saved). Returns { stdout, stderr, error? }."""
    try:
        return script_service.run_script(
            body.path, body.file, body.script, field_mapping=_mapping_dict(body.field_mapping)
        )
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.get("/scripts", summary="List saved dataset scripts")
def list_dataset_scripts():
    try:
        return {"scripts": script_service.list_saved_scripts()}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.get("/script/{script_id:path}", summary="Get saved script body")
def get_dataset_script(script_id: str):
    try:
        body = script_service.get_saved_script(script_id)
        if body is None:
            raise HTTPException(404, "Script not found")
        return {"id": script_id, "body": body}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.post("/script/save", summary="Save script by name")
def save_dataset_script(body: SaveScriptBody):
    """Body: { id: string, body: string }. Saves as .ai-workspace/dataset_scripts/{id}.py"""
    try:
        sid = body.id or ""
        script_service.save_script(sid, body.body or "")
        return {"ok": True}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))


@router.delete("/script/{script_id:path}", summary="Delete saved script")
def delete_dataset_script(script_id: str):
    try:
        ok = script_service.delete_script(script_id)
        return {"deleted": ok}
    except PathOutsideWorkspaceError as e:
        raise HTTPException(403, str(e.message))
